### PR TITLE
[MLIR] Add f6E3M2FN type

### PR DESCRIPTION
**Pull request by sergey-kozub**
This PR adds `f6E3M2FN` type to mlir.

`f6E3M2FN` type is proposed in [OpenCompute MX Specification](https://www.opencompute.org/documents/ocp-microscaling-formats-mx-v1-0-spec-final-pdf). It defines a 6-bit floating point number with bit layout S1E3M2. Unlike IEEE-754 types, there are no infinity or NaN values.

```c
f6E3M2FN
- Exponent bias: 3
- Maximum stored exponent value: 7 (binary 111)
- Maximum unbiased exponent value: 7 - 3 = 4
- Minimum stored exponent value: 1 (binary 001)
- Minimum unbiased exponent value: 1 − 3 = −2
- Has Positive and Negative zero
- Doesn't have infinity
- Doesn't have NaNs

Additional details:
- Zeros (+/-): S.000.00
- Max normal number: S.111.11 = ±2^(4) x (1 + 0.75) = ±28
- Min normal number: S.001.00 = ±2^(-2) = ±0.25
- Max subnormal number: S.000.11 = ±2^(-2) x 0.75 = ±0.1875
- Min subnormal number: S.000.01 = ±2^(-2) x 0.25 = ±0.0625
```

Related PRs:
- [PR-94735](https://github.com/llvm/llvm-project/pull/94735) [APFloat] Add APFloat support for FP6 data types
- [PR-97118](https://github.com/llvm/llvm-project/pull/97118) [MLIR] Add f8E4M3 type - was used as a template for this PR

### CODE REVIEW

**Comment by sergey-kozub** [comment]
I'm planning to add a similar PR for f6E2M3FN type (also has APFloat semantics already).

File: llvm/unittests/ADT/APFloatTest.cpp [hunk 1]
```diff
@@ -2084,16 +2084,17 @@ TEST(APFloatTest, getSmallestNormalized) {
...
-  expected = APFloat(APFloat::Float4E2M1FN(), "0x1p0");
   EXPECT_FALSE(test.isNegative());
   EXPECT_TRUE(test.isFiniteNonZero());
   EXPECT_FALSE(test.isDenormal());
   EXPECT_TRUE(test.bitwiseIsEqual(expected));
   EXPECT_TRUE(test.isSmallestNormalized());
+
+  test = APFloat::getSmallestNormalized(APFloat::Float4E2M1FN(), false);
+  expected = APFloat(APFloat::Float4E2M1FN(), "0x1p0");
   EXPECT_FALSE(test.isNegative());
```
Comment (durga4github) [inline llvm/unittests/ADT/APFloatTest.cpp:2098]: good catch, Thanks for fixing this!

**Comment by durga4github** [comment]
The changes look good to me,

@ThomasRaoux , Could you please help with a review?

File: mlir/include/mlir/IR/Builders.h [hunk 2]
```diff
@@ -67,6 +67,7 @@ class Builder {
   FloatType getFloat8E4M3FNUZType();
   FloatType getFloat8E4M3B11FNUZType();
   FloatType getFloat8E3M4Type();
+  FloatType getFloat6E3M2FNType();
```
Comment (dcaballe) [inline mlir/include/mlir/IR/Builders.h:70]: nit: it looks like methods are ordered by type bitwidth (?) so I would move this one before the Float8 ones

File: mlir/include/mlir/IR/BuiltinTypes.h [hunk 3]
```diff
@@ -415,9 +416,9 @@ inline bool BaseMemRefType::isValidElementType(Type type) {
 inline bool FloatType::classof(Type type) {
   return llvm::isa<Float8E5M2Type, Float8E4M3Type, Float8E4M3FNType,
                    Float8E5M2FNUZType, Float8E4M3FNUZType,
-                   Float8E4M3B11FNUZType, Float8E3M4Type, BFloat16Type,
-                   Float16Type, FloatTF32Type, Float32Type, Float64Type,
-                   Float80Type, Float128Type>(type);
+                   Float8E4M3B11FNUZType, Float8E3M4Type, Float6E3M2FNType,
```
Comment (dcaballe) [inline mlir/include/mlir/IR/BuiltinTypes.h:419]: Same? (follow bitwidth order)

File: mlir/include/mlir/IR/BuiltinTypes.td [hunk 4]
```diff
@@ -233,6 +233,24 @@ def Builtin_Float8E3M4 : Builtin_FloatType<"Float8E3M4", "f8E3M4"> {
...
 
+//===----------------------------------------------------------------------===//
+// Float6E3M2FNType
+
+def Builtin_Float6E3M2FN : Builtin_FloatType<"Float6E3M2FN", "f6E3M2FN"> {
+  let summary = "6-bit floating point with 3 bits exponent and 2 bit mantissa";
+  let description = [{
+    An 6-bit floating point type with 1 sign bit, 3 bits exponent and 2 bits
+    mantissa. This is not a standard type as defined by IEEE-754, but it
+    follows similar conventions with the following characteristics:
```
Comment (dcaballe) [inline mlir/include/mlir/IR/BuiltinTypes.td:244]: Should we add a reference to the OCP spec or at least mention it?

File: mlir/lib/Dialect/Arith/Transforms/EmulateUnsupportedFloats.cpp [hunk 5]
```diff
@@ -61,6 +61,7 @@ static std::optional<FloatType> parseFloatType(MLIRContext *ctx,
       .Case("f8E5M2FNUZ", b.getFloat8E5M2FNUZType())
       .Case("f8E4M3FNUZ", b.getFloat8E4M3FNUZType())
       .Case("f8E3M4", b.getFloat8E3M4Type())
+      .Case("f6E3M2FN", b.getFloat6E3M2FNType())
```
Comment (dcaballe) [inline mlir/lib/Dialect/Arith/Transforms/EmulateUnsupportedFloats.cpp:64]: Same

File: mlir/lib/IR/BuiltinTypes.cpp [hunk 6]
```diff
@@ -91,6 +91,8 @@ IntegerType IntegerType::scaleElementBitwidth(unsigned scale) {
 //===----------------------------------------------------------------------===//
 
 unsigned FloatType::getWidth() {
+  if (llvm::isa<Float6E3M2FNType>(*this))
+    return 6;
```
Comment (dcaballe) [inline mlir/lib/IR/BuiltinTypes.cpp:95]: Why don't we use APFloat here?

**Review by dcaballe — COMMENTED** [review]
LG in general. Dropped some minor comments

**Comment by stellaraccident** [comment]
Just curious as there have been various conversations: have we found a use case for the scalar form of f6 yet? When I've drilled in before, it seemed like there might be a case for that but all uses in the wild I could see were for the shared exponent forms (which are definitely not a scalar type).

File: mlir/include/mlir/IR/Builders.h (hunk 2 above)
Comment (sergey-kozub) [reply mlir/include/mlir/IR/Builders.h]: Done, here and in the other places.

File: mlir/include/mlir/IR/BuiltinTypes.h (hunk 3 above)
Comment (sergey-kozub) [reply mlir/include/mlir/IR/BuiltinTypes.h]: Done.

File: mlir/include/mlir/IR/BuiltinTypes.td (hunk 4 above)
Comment (sergey-kozub) [reply mlir/include/mlir/IR/BuiltinTypes.td:244]: Added a link.

File: mlir/lib/Dialect/Arith/Transforms/EmulateUnsupportedFloats.cpp (hunk 5 above)
Comment (sergey-kozub) [reply mlir/lib/Dialect/Arith/Transforms/EmulateUnsupportedFloats.cpp]: Done.

File: mlir/lib/IR/BuiltinTypes.cpp (hunk 6 above)
Comment (sergey-kozub) [reply mlir/lib/IR/BuiltinTypes.cpp:95]: I'll do this in a separate PR, if you don't mind.

Comment (sergey-kozub) [reply mlir/lib/IR/BuiltinTypes.cpp:95]: Created https://github.com/llvm/llvm-project/pull/107372

**Comment by sergey-kozub** [comment]
> have we found a use case for the elemental form of f6 yet?

I'm not sure about that. This type is supposed to be used as a block scaled format, i.e. for reducing the weights size in forward layers.

**Comment by sergey-kozub** [comment]
The PR appears green now, should I press the "Squash and merge" button, or should I let a reviewer do it instead?

**Review by apivovarov — APPROVED** [review]

**Comment by apivovarov** [comment]
> The PR appears green now, should I press the "Squash and merge" button, or should I let a reviewer do it instead?

If github allows you to merge - you can merge.

**Comment by dcaballe** [comment]
> Just curious as there have been various conversations: have we found a use case for the elemental form of f6 yet? When I've drilled in before, it seemed like there might be a case for that but all uses in the wild I could see were for the shared scale forms (which are definitely not a scalar type).

I believe the goal is to use the element types as a building block to model block-scaled types. After looking at how block-scaled types are modeled in other projects, I reached the conclusion that we need to build more expertise before introducing the full concept of a block-scaled type in MLIR, if at all. By using element types, we could experiment with block-scaled types by modeling them with separate tensors: one for all the block elements and another for all the block scales. Regardless of our final approach, we would need this basic support in `APFloat` to handle their specific semantics.

**Review by dcaballe — APPROVED** [review]

**Pull request merged**